Add internal AppExecLink support to link APIs

carlossanlop opened the pull request:
This is a continuation of https://github.com/dotnet/runtime/pull/57996 

AppExecLinks, also known as Application Execution Aliases, are a Windows-specific kind of symbolic link for Windows Apps. It's a scenario we should support when resolving links with our FileSystemInfo APIs or the static File/Directory APIs.

AppExecLinks are an [explicitly supported scenario](https://github.com/PowerShell/PowerShell/blob/56d22bc3865fca445145fd685d6d6f6d63194701/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L8509) by PowerShell, along with mount points. Our APIs should handle these types of links as well, to facilitate the task of the PowerShell team to port their code to use these new link APIs.

carlossanlop committed:
Add internal AppExecLink support to link APIs

carlossanlop committed:
PlatformDetection property to use as condition if Windows machine has an AppExecLinks directory.

carlossanlop committed:
AppExecLink unit test.

carlossanlop committed:
Address suggestions.

iSazonov commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -539,13 +539,59 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
                     Debug.Assert(!PathInternal.IsPartiallyQualified(targetPath));
                     return targetPath.ToString();
                 }
+                else if (rbSymlink.ReparseTag == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_APPEXECLINK)
```
> how is this one special?

It is _unnormal_ reparse point :-) so we had to add it explicitly in PowerShell.

In common, reparse point list is "open" and if new reparse points will be added in future _perhaps_ we will have to add its explicitly too.

I'd remind about OneDrive. .Net API should be tested for OneDrive paths/links too (at list manually).

carlossanlop replied:
I manually tested enumeration of the OneDrive directory (ensuring there were cloud files) and the behavior was the same between 5.0 and in 6.0: I can read the file name and the file attributes, and the files are not getting unexpectedly downloaded - their cloud icon stays next to the files when viewing them in the File Explorer.

@isazonov do you know any specific scenarios in which the OneDrive files get unexpectedly downloaded when manipulating them with .NET?

iSazonov replied:
@carlossanlop These links are related OneDrive and could be interesting for you:
https://github.com/PowerShell/PowerShell/pull/9895
https://github.com/PowerShell/PowerShell/pull/9509
https://github.com/PowerShell/PowerShell/pull/8745

Also for removing https://github.com/PowerShell/PowerShell/pull/15571

Feel free to ask me if you need additional comments.

iSazonov commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -539,6 +539,43 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
+                    Debug.Assert(success);
+
+                    // The target file is at index 2
+                    if (rbAppExecLink.StringCount >= 3)
```
I did not found docs for IO_REPARSE_TAG_APPEXECLINK in time we add the code in PowerShell.

carlossanlop replied:
We're starting a conversation with the Windows team about this, since the lack of docs is a blocker for us.

iSazonov replied:
PowerShell team found the const in Windows SDK so it is public and the fact opened a way to use the const.

carlossanlop replied:
@iSazonov can you share a link to the public Windows SDK location where you found it?

I'm not sure such a link would qualify as "public documentation" though.

iSazonov replied:
@carlossanlop If I remember right @SteveL-MSFT found this in Windows SDK _files_ (I guess in *.cpp).

SteveL-MSFT replied:
I don't remember where exactly I got the original buffer struct from, but given the name, I think most likely from the SO post.  Since I can't find any public docs on the struct, it should probably not be relied upon as it could change.  I'll start an internal thread with the AppX team to see about getting it documented publicly.

iSazonov replied:
I found where we discuss this https://github.com/PowerShell/PowerShell/pull/10331#discussion_r316528849 - there is a link to official Microsoft example code ^-)

iSazonov commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -539,13 +539,59 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
+                        count++;
+                        if (count == 2)
+                        {
+                            start = i + 1; // exclude null char
```
Can we use `public static unsafe int IndexOf(ref char searchSpace, char value, int length)` from internal SpanHelpers?

carlossanlop replied:
> if the buffer is corrupt such that \0 is the last char

If the AppExecLink struct was malformed from the beginning, then either the `StringCount` was incorrect, and/or the `sizeof` the struct was incorrect. I could add a check to ensure the `Slice` is not going beyond the `sizeof` the struct.

danmoseley replied:
We do different things based on how confident we are in the blob I guess. Eg., for reading stat etc files, we will throw if the file is malformed. I guess the min bar is to be memory safe - and Span is safe, so I think it's fine as is unless and until we discover otherwise. Either way is fine.

carlossanlop commented:
Had a conversation with Dan about this issue and it doesn't meet the bar for a 6.0 backport.

carlossanlop commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -289,6 +291,44 @@ public static bool IsSubstAvailable
+                        }
+                    }
+                }
+                catch { }
```
That's a valid concern. I guess we shouldn't expect these enumeration folders to throw with the way the are currently written. I'll remove the try catch and make sure the checks are done carefully.

carlossanlop commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -289,6 +291,44 @@ public static bool IsSubstAvailable
+                        string windowsAppsDir = Path.Join(localAppData, "Microsoft", "WindowsApps");
+                        if (Directory.Exists(windowsAppsDir))
+                        {
+                            // return Directory.GetFiles(windowsAppsDir, "*.exe", new EnumerationOptions() { RecurseSubdirectories = true, MaxRecursionDepth = 3 }).Length > 0;
```
Removing this comment.

carlossanlop committed:
Remove big try catch in PlatformDetection property.

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -572,7 +619,7 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
                 // We need to fallback to manually traverse the links and return the target of the last resolved link.
                 int error = Marshal.GetLastWin32Error();
-                if (IsPathUnreachableError(error))
+                if (IsPathUnreachableError(error) || error == Interop.Errors.ERROR_CANT_ACCESS_FILE /* Possibly broken AppExecLink */)
```
The fact that the `IsPathUnreachableError` method is called in other places (specifically, the piece of code you shared), is the reason why I am placing the `ERROR_CANT_ACCESS_FILE` here, and not inside that method.

But the name of the method is to determine if a path is unreachable, so you're correct, I think it makes sense to catch `ERROR_CANT_ACCESS_FILE` there as well. I will try to add a unit test for `FileSystem.Attributes.Windows` that generates this error and catches it (hopefully, this can happen with an AppExecLink now).

I'll add the comment too.

carlossanlop committed:
Fix csproj alphabetized ordering.

carlossanlop committed:
Extra checks in for loop to find null chars of appexeclink stringlist.
Add CANT_ACCESS_FILE inside IsPathUnreachableError. Add comment to that method to avoid catching ACCESS_DENIED, SHARING_VIOLATION, SEM_TIMEOUT.

carlossanlop commented:
> This is presumably a Windows 8 or Windows 10 feature. Will there be any problems on Windows 7? I assume not, it just won't hit any.

Execution Aliases for UWP applications were introduced in Windows 10 Fall Creators Update (1709/RS3, build 16299). Windows versions before that should not have appexeclink files in their system. If they do, the scenario would be unsupported on the Windows side, so either `OpenSafeFileHandle` or `DeviceIoControl` would fail and we would handle the error.

iSazonov commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.REPARSE_DATA_BUFFER.cs:
```diff
@@ -38,5 +38,14 @@ internal struct MountPointReparseBuffer
             public ushort PrintNameLength;
         }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct AppExecLinkReparseBuffer
+        {
+            public uint ReparseTag;
+            public ushort ReparseDataLength;
+            public ushort Reserved;
+            public uint StringCount;
+        }
```
It is not related the PR but I wonder why do such structs returned from P/Invokes would not be _readonly_ structs?

iSazonov commented on src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:
```diff
@@ -132,6 +134,7 @@ internal static bool IsPathUnreachableError(int errorCode)
                 case Interop.Errors.ERROR_NETWORK_ACCESS_DENIED:
                 case Interop.Errors.ERROR_INVALID_HANDLE:           // eg from \\.\CON
                 case Interop.Errors.ERROR_FILENAME_EXCED_RANGE:     // Path is too long
+                case Interop.Errors.ERROR_CANT_ACCESS_FILE:         // Broken AppExecLink files may cause this
```
No false positives could be for the error in the context?

carlossanlop replied:
I'm not sure I understand the question. Would you mind rephrasing that?

iSazonov replied:
> Broken AppExecLink files may cause this

My concern is about this comment. Could there be other files that cause the same error?

iSazonov commented on src/libraries/System.IO.FileSystem/tests/AppExecLinks.Windows.cs:
```diff
@@ -0,0 +1,57 @@
+    [PlatformSpecific(TestPlatforms.Windows)]
+    public class AppExecLinks : BaseSymbolicLinks
+    {
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.HasUsableAppExecLinksDirectory))]
```
Sorry if I skipped this - will this throw if HasUsableAppExecLinksDirectory return false or the test could be skipped always silently in CIs?

carlossanlop replied:
It should return false and cause the test to be skipped due to the absence of a WindowsApps dir.

The property this attribute is pointing to, should not throw. If it throws, it's a bug and needs to be fixed.

iSazonov replied:
My concern is that in the case the tests can be never really evaluated on CIs.

adamsitnik commented:
> Had a conversation with Dan about this issue and it doesn't meet the bar for a 6.0 backport.

The bar for approving changes for .NET 6 is:

> 1.	Broken or significantly incomplete 6.0 scenario

I would not say that it's significantly incomplete, but since it blocks one of our 1st party customers from switching to the new APIs (https://github.com/PowerShell/PowerShell/pull/15960/) we should rather include it in 6.0. Just my two cents.

danmoseley commented:
> I would not say that it's significantly incomplete, but since it blocks one of our 1st party customers from switching to the new APIs (PowerShell/PowerShell#15960) we should rather include it in 6.0. Just my two cents.

It does, but I figured this customer already has code that works fine. So it would really just be code cleanup for them. If there were other customers that want to adopt the API and are missing support for "AppExecLink" that would be a little different.

That being said, @iSazonov if we took this change, would it be the last piece allowing significant code cleanup from Powershell? Would Powershell be able to adopt it promptly? Is that a signficant win for you guys?

iSazonov commented:
> That being said, @iSazonov if we took this change, would it be the last piece allowing significant code cleanup from Powershell? Would Powershell be able to adopt it promptly? Is that a signficant win for you guys?

For me, it is not critical to have the API today since we are near RC too (and it is LTS version too!) and I guess PowerShell team doesn't want to catch a regression or bug. It is better to ask @SteveL-MSFT directly.

Really there are some API proposal PowerShell can benefit from (specially path enumeration API) and I hope to see them implemented in next milestone early so that we can migrate and test early too.

carlossanlop commented:
We had a conversation with the Windows team that owns AppExecLinks. They explained that we are not supposed to care about the target path. AppExecLinks are meant to work seamlessly as any other executable. Hence why they never exposed the APIs that we were adding in this PR, and which PowerShell is also calling.

Having learned that, we decided won't be treating AppExecLinks like symlinks and junctions. We will still return true for `Attributes.HasFlag(FileAttributes.ReparsePoint)`, but the new APIs `LinkTarget` and `ResolveLinkTarget` will return `null` (this scenario already works, thanks to @Jozkee 's [PR](https://github.com/dotnet/runtime/pull/58391)).

We also had a conversation with @SteveL-MSFT , @rjmholt, @daxian-dbw from the PowerShell team to let them know what the Windows people told us, so they make adjustments to their code as needed.

With that said, since we are already providing support to symlinks, [junctions](https://github.com/dotnet/runtime/pull/57996) and appexeclinks as expected, the PowerShell scenario will be complete in 6.0 after we merge [this backport to RC2](https://github.com/dotnet/runtime/pull/58399) (@danmoseley).

@iSazonov
